packages/esm-patient-list-app/src/offline/offlineData.ts

```diff
@@ -0,0 +1,239 @@
+import { createGlobalStore, getCurrentUser, getGlobalStore } from '@openmrs/esm-framework';
+import Dexie from 'dexie';
+import { LoadState, OfflinePatient } from './types';
+
+/**
+ * Thoughts on behaviour when LOADING or RELOADING are present on DB construction
+ * This state means, that the given resource was being loaded but the process was interrupted
+ * There are two possible ways to handle this:
+ * - set LOADING to LOADING_ERROR and RELOADING to RELOADING_ERROR and let the user trigger the loads again
+ * - retrigger the handlers silently on construction
+ */
+
+/**
+ * required interfaces
+ * for handlers:
+ * - adding patient handlers
+ *
+ * for consumers (components/patient list):
+ * - adding a patient
+ * - removing a patient
+ * - reloading a patient
+ * - subscribing to the list of patients
+ * - retrieving the list of patients
+ * - retrieving the list of handlers (should not change after the setup)
+ */
+
+interface PatientHandlerObject {
+  onLoadPatient: (patientUuid: string, onSuccess: () => void, onError: (err: Error) => void) => void;
+  onRemovePatient: (patientUuid: string) => void;
+}
+
+/**
+ * stores handlers that want to be called when a patient is added or removed
+ */
+const patientStore = createGlobalStore<Record<string, PatientHandlerObject>>('offline-patients-handler', {});
+export function addPatientHandler(uuid: string, handlerObject: PatientHandlerObject) {
+  getGlobalStore<Record<string, PatientHandlerObject>>('offline-patients-handler').setState((state) => ({
+    ...state,
+    [uuid]: handlerObject,
+  }));
+}
+
+function getOfflineHandlers(): Array<string> {
+  return Object.keys(patientStore.getState());
+}
+
+// Declare Database
+class OfflinePatientDatabase extends Dexie {
+  private offlinePatients: Dexie.Table<OfflinePatient, string>; // id is number in this case
+  private currentUser = 'unauthorized_user'; // will be automatically set to current user if logged in
+  private changeSubscribers = new Set<(data: Array<OfflinePatient>) => void>();
+  private broadcastChange = () => {}; // is overwritten if the Broadcastchannel is available
+
+  constructor() {
+    super('OfflinePatientDatabase');
+    this.version(5).stores({
+      offlinePatients: 'uuid',
+    });
+    this.offlinePatients = this.table('offlinePatients');
+    getCurrentUser().subscribe((user) => {
```

Comment:
Hm let's avoid using this. It's currently being refactored and rxjs is not used (for this) any more.

Reply:
What would be the way to get user/userChanges then? Or can we assume, that a user change triggers a application reload?